### PR TITLE
Redesign the interface to the unikernel HermitCore

### PR DESCRIPTION
**Pull request by stlankes**
We are developing the unikernel HermitCore, where the kernel is written in Rust and is already part of the Rust Standard Library. The interface between the standard library and the kernel based on a small C library. With this pull request, we remove completely the dependency to C and use lld as linker. Currently, the kernel will be linked to the application as static library, which is published at https://github.com/hermitcore/libhermit-rs.

We don’t longer support the C interface to the kernel. Consequently, we remove this part from the Rust Standard Library.

### CODE REVIEW

**Comment by JohnCSimon** [comment]
Ping from triage 
@varkor @withoutboats Can you please review this PR?
thanks.

**Comment by stlankes** [comment]
Yes, maybe I should explain our motivation. I am part of a research group for system software at a german university and use Rust in my [operating system classes](https://github.com/RWTH-OS/eduOS-rs). We want to maintain and extend this part of libstd.

The unikernel HermitCore is one of our projects. Unikernels are specialized, single-address-space machine images. Such kernels are typically used in the area of cloud computing. They shrink the attack surface and resource footprint of cloud services. The best known unikernel is maybe [MirageOS](https://mirage.io). We already developed a version of HermitCore, which is part of libstd. However, it is mixture of C and Rust. Our objective is to remove C completely from our toolchain. C is in our opinion not valid to develop secure applications.

The current version is only working within our hypervisor [uhyve](https://crates.io/crates/uhyve). A prototype is already working with @phil-opp [bootimage](https://github.com/rust-osdev/bootimage), which converts a kernel into a valid boot image. Consequently, it runs on common hypervisors like qemu. A network is currently missing. We have also a prototype, which uses smoltcp as  network stack. But it isn’t in a state to publish it. 

So, we have something in our pipeline. The current patch is already huge. This is our motivation to publish the current version and smaller extensions later.

I hope that I have explained our motivation well.

**Comment by alexcrichton** [comment]
Ok thanks for the info!

As I mentioned before we don't really have strict policies right now one way or another about how to organize libstd. It's generally left up to the reviewer and a few defacto ways that we manage a separate target in libstd. For example:

* You'll be the owner of this directory. You'll be cc'd if something looks wonky or if assistance is needed.
* No code review is guaranteed to happen for any code within this hermit-specific directory. Integration points with the rest of libstd will be reviewed, but hermit-specific code is not guaranteed to receive review (e.g. I wouldn't review anything platform-specific in this PR).
* Changes to libstd will not be gated on the hermit target, so this will likely break over time as libstd is refactored, tweaked, updated, etc. You won't always get warning to this happening as well, it'll just break in nightlies.
* We're in the works of developing a more fleshed out policy for target support, and that may change the calculus here, although it's unknown by how much. Just want to make sure you're aware!

If that all sounds good, I'll review the integration points here and can r+

**Comment by stlankes** [comment]
It is fine for me!

File: src/libstd/rt.rs [hunk 1]
```diff
@@ -59,7 +59,6 @@ fn lang_start_internal(main: &(dyn Fn() -> i32 + Sync + crate::panic::RefUnwindS
 #[cfg(not(test))]
 #[lang = "start"]
 fn lang_start<T: crate::process::Termination + 'static>
-    (main: fn() -> T, argc: isize, argv: *const *const u8) -> isize
-{
+    (main: fn() -> T, argc: isize, argv: *const *const u8) -> isize {
```
Comment (alexcrichton) [inline src/libstd/rt.rs]: Stray change?

File: src/libunwind/build.rs [hunk 2]
```diff
@@ -43,6 +43,8 @@ fn main() {
         println!("cargo:rustc-link-lib=gcc_s");
     } else if target.contains("redox") {
         // redox is handled in lib.rs
+    } else if target.contains("hermit") {
+        //println!("cargo:rustc-link-lib=unwind");
```
Comment (alexcrichton) [inline src/libunwind/build.rs]: No longer needed perhaps?

File: src/libunwind/libunwind.rs [hunk 3]
```diff
@@ -1,7 +1,26 @@
...
+#[repr(u8)]
+#[allow(missing_copy_implementations)]
+#[allow(missing_debug_implementations)]
+pub enum c_void {
+    // Two dummy variants so the #[repr] attribute can be used.
+    #[doc(hidden)]
+    __variant1,
+    #[doc(hidden)]
+    __variant2,
+}
```
Comment (alexcrichton) [inline src/libunwind/libunwind.rs]: Would it be possible to move these implementations to `libc`?

File: src/libunwind/build.rs (hunk 2 above)
Comment (stlankes) [reply src/libunwind/build.rs]: Yes, I have to remove it

File: src/libunwind/libunwind.rs (hunk 3 above)
Comment (stlankes) [reply src/libunwind/libunwind.rs]: It is already in libc. I forgot only to remove my workaround. I will fix it.

**Comment by stlankes** [comment]
@alexcrichton I follow your suggestions and revise my code.

**Comment by stlankes** [comment]
@alexcrichton Okay, I guess I was naive.  I moved the C interface to special crate to simplify the maintenance. The crate isn't part of the whitelist. Do you add it on the whitelist or should I revert the last commit?

**Comment by alexcrichton** [comment]
Yes adding that to the whitelist should be fine, most new targets in Rust nowadays end up taking this strategy of moving the very low-level support to a separate crate that libstd then uses.

**Comment by alexcrichton** [comment]
@stlankes hm was that last commit pushed here by accident perhaps?

**Comment by stlankes** [comment]
The website shows a conflict in `src/libtest/lib.rs` and I fixed it. But I think that I never changed `src/libtest/lib.rs`.

**Comment by stlankes** [comment]
@alexcrichton What does test failed?

**Comment by JohnTitor** [comment]
It's a network error.
@bors retry

**Comment by stlankes** [comment]
@alexcrichton Naive question. I need for my toolchain `llvm-tools-preview` (e,g. `llvm-objcopy`). I set in `config.toml` the variable `llvm-tools` to `true`. After building the toolchain, I do not find `llvm-objcopy`. Do you have a hint for me?

**Comment by stlankes** [comment]
@alexcrichton No problem!

**Comment by stlankes** [comment]
@alexcrichton I have a [version of RustyHermit](https://github.com/hermitcore/hermit-abi/blob/with-hermit/Cargo.toml#L20), where I include my libos as crate instead of a static library. In this case, I am able to call directly Rust code from my kernel (e.g [https://github.com/hermitcore/rust/blob/with-hermit/src/libstd/sys/hermit/mutex.rs#L12](https://github.com/hermitcore/rust/blob/with-hermit/src/libstd/sys/hermit/mutex.rs#L12)).

However, I have to modify a few helper creates (log, rust-x86, rust-cupid,...). For instance, I modified [log](https://github.com/hermitcore/log/commit/3ffa4914ecfb048edd46f2255a6acc87341c69dd).  Do you think that this a good idea? Can this be part of libstd?

**Comment by alexcrichton** [comment]
I think we'll want to continue to be pretty conservative with dependencies of the standard library, so we don't have a great way of adding new dependencies to libstd for third-party targets at this time unfortunately.

**Comment by stlankes** [comment]
No problem. I was expecting the answer....

**Pull request merged**